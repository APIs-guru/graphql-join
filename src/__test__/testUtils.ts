declare var expect;

import {
  Source,
  DocumentNode,
  GraphQLError,
  GraphQLSchema,
  GraphQLFieldResolver,
  GraphQLResolveInfo,
  ExecutionResult,

  graphql,
  parse,
  print,
  buildSchema,
  printSchema,
  formatError,
} from 'graphql';
import * as _ from 'lodash';

import { joinSchemas, ProxyContext } from '../index';
import { stubSchema } from '../utils';

expect.addSnapshotSerializer({
  print(val) {
    return typeof val === 'string' ?
      val.trim() :
      JSON.stringify(val, null, 2);
  },
  test(val) {
    return true;
  },
});

function fakeFieldResolver(schemaName: string): GraphQLFieldResolver<any,any> {
  return (
    _1, _2, _3,
    {fieldName, parentType}: GraphQLResolveInfo
  ) => {
    return `${schemaName}::${parentType.name}::${fieldName}`;
  }
}

function makeProxy(schemaName: string, schema: GraphQLSchema) {
  return (queryAST: DocumentNode) => {
    const query = print(queryAST);
    expect(query).toMatchSnapshot();
    return graphql({
      schema,
      source: query,
      fieldResolver: fakeFieldResolver(schemaName),
    });
  };
}

type TestSchema = string;
type TestSchemasMap = { [name: string]: TestSchema };
export function testJoin(testSchemas: TestSchemasMap, joinSDL: string) {
  const remoteSchemas = _.mapValues(testSchemas, (sdl, name) => {
    const schema = buildSchema(new Source(sdl, name));
    stubSchema(schema);
    return { schema };
  });

  const joinAST = parse(new Source(joinSDL, 'Join SDL'));
  const schema = joinSchemas(joinAST, remoteSchemas);

  expect(schema).toBeInstanceOf(GraphQLSchema);
  expect(printSchema(schema)).toMatchSnapshot();
  return async (
    query: string,
    results?: { [schemaName: string]: ExecutionResult }
  ) => {
    const proxyFns = _.mapValues(remoteSchemas, ({schema}, name) => {
      if (results && results[name]) {
        return (() => Promise.resolve(results[name]));
      }
      return makeProxy(name, schema)
    });

    const result = await graphql({
      schema,
      source: new Source(query, 'ClientQuery'),
      contextValue: new ProxyContext(proxyFns),
    });
    expect([
      query,
      resultToJSON(result),
    ]).toMatchSnapshot();
  };
}

function resultToJSON(result) {
  return {
    ...result,
    errors: (result.errors || []).map(formatError),
  };
}

