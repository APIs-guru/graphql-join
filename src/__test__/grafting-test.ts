declare var describe, test;

import { testJoin } from './testUtils';

describe('grafting tests', () => {
  test('extend Query type', async () => {
    const execute = testJoin({
      test1: `type Query { foo: String }`,
      test2: `
        schema { query: Test2Query }
        type Test2Query { bar: Bar }
        type Bar { baz: String }
      `
    },`
      schema {
        query: Query # test1 Query
      }
      extend type Query {
        bar: Bar @resolveWith(query: "bar")
      }
      query bar @send(to: "test2") {
        bar { ...CLIENT_SELECTION }
      }
    `);
    await execute('{ foo }');
    await execute('{ bar { baz } }');
    await execute('{ foo bar { baz } }');
  });
});
