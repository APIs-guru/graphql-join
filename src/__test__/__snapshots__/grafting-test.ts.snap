// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`grafting tests extend Query type 1`] = `
type Bar {
  baz: String
}

type Query {
  foo: String
  bar: Bar
}
`;

exports[`grafting tests extend Query type 2`] = `
{
  foo
}
`;

exports[`grafting tests extend Query type 3`] = `
[
  "{ foo }",
  {
    "data": {
      "foo": "test1::Query::foo"
    },
    "errors": []
  }
]
`;

exports[`grafting tests extend Query type 4`] = `
{
  bar {
    baz
  }
}
`;

exports[`grafting tests extend Query type 5`] = `
[
  "{ bar { baz } }",
  {
    "data": {
      "bar": {
        "baz": "test2::Bar::baz"
      }
    },
    "errors": []
  }
]
`;

exports[`grafting tests extend Query type 6`] = `
{
  foo
}
`;

exports[`grafting tests extend Query type 7`] = `
{
  bar {
    baz
  }
}
`;

exports[`grafting tests extend Query type 8`] = `
[
  "{ foo bar { baz } }",
  {
    "data": {
      "foo": "test1::Query::foo",
      "bar": {
        "baz": "test2::Bar::baz"
      }
    },
    "errors": []
  }
]
`;
